drivers/gpio/gpio_imx.c

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright (c) 2018, NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <device.h>
+#include <gpio.h>
+#include <soc.h>
+#include <gpio_imx.h>
+
+#include "gpio_utils.h"
+
+struct imx_gpio_config {
+	GPIO_Type *base;
+};
+
+struct imx_gpio_data {
+	/* port ISR callback routine address */
+	sys_slist_t callbacks;
+	/* pin callback routine enable flags, by pin number */
+	u32_t pin_callback_enables;
+};
+
+static int imx_gpio_configure(struct device *dev,
+			       int access_op, u32_t pin, int flags)
+{
+	const struct imx_gpio_config *config = dev->config->config_info;
+	gpio_init_config_t pin_config;
+	u32_t i;
+
+	/* Check for an invalid pin configuration */
+	if ((flags & GPIO_INT) && (flags & GPIO_DIR_OUT)) {
+		return -EINVAL;
+	}
+
+	pin_config.direction = ((flags & GPIO_DIR_MASK) == GPIO_DIR_IN)
+		? gpioDigitalInput : gpioDigitalOutput;
+
+	if (flags & GPIO_INT) {
+		if (flags & GPIO_INT_EDGE) {
+			if (flags & GPIO_INT_ACTIVE_HIGH) {
+				pin_config.interruptMode = gpioIntRisingEdge;
+			} else if (flags & GPIO_INT_DOUBLE_EDGE) {
+				pin_config.interruptMode =
+					gpioIntRisingOrFallingEdge;
+			} else {
+				pin_config.interruptMode =
+					gpioIntFallingEdge;
+			}
+		} else { /* GPIO_INT_LEVEL */
+			if (flags & GPIO_INT_ACTIVE_HIGH) {
+				pin_config.interruptMode = gpioIntHighLevel;
+			} else {
+				pin_config.interruptMode = gpioIntLowLevel;
+			}
+		}
+	} else {
+		pin_config.interruptMode = gpioNoIntmode;
+	}
+
+	if (access_op == GPIO_ACCESS_BY_PIN) {
+		pin_config.pin = pin;
+		GPIO_Init(config->base, &pin_config);
+	} else {	/* GPIO_ACCESS_BY_PORT */
+		for (i = 0; i < 32; i++) {
+			pin_config.pin = pin;
+			GPIO_Init(config->base, &pin_config);
+		}
+	}
+
+	return 0;
+}
+
+static int imx_gpio_write(struct device *dev,
+			   int access_op, u32_t pin, u32_t value)
+{
+	const struct imx_gpio_config *config = dev->config->config_info;
+
+	if (access_op == GPIO_ACCESS_BY_PIN) {
+		GPIO_WritePinOutput(config->base, pin, value);
+	} else { /* GPIO_ACCESS_BY_PORT */
+		GPIO_WritePortOutput(config->base, value);
+	}
+
+	return 0;
+}
+
+static int imx_gpio_read(struct device *dev,
+			  int access_op, u32_t pin, u32_t *value)
+{
+	const struct imx_gpio_config *config = dev->config->config_info;
+
+	if (access_op == GPIO_ACCESS_BY_PIN) {
+		*value = GPIO_ReadPinOutput(config->base, pin);
```

Comment:
Shouldn't be GPIO_ReadPinInput() used here? Anyway GPIO_ReadPinInput() and GPIO_ReadPinOutput() does the same thing in BSP driver ext/hal/nxp/imx/drivers/gpio_imx.h.

Reply:
OK